### PR TITLE
current state of accelerated GJK

### PR DESCRIPTION
**Pull request by MaartenBehn**
- I merged your performance optimizations. 
- Added benchmarks
- Changed to_dict functions for json dumping.
- Fixed Unit tests

If you want to further tinker with the performance of GJK Nasterov please use this state of the code for easy merge later.

### CODE REVIEW

**Comment by MaartenBehn** [comment]
> can we close #67 and #70?

yeah sure

File: distance3d/gjk/_gjk_original.py [hunk 1]
```diff
@@ -5,7 +5,11 @@
 EPSILON = 10.0 * np.finfo(float).eps
 
 
-def gjk_distance_original(collider1, collider2):
+def gjk_distance_iterations(collider1, collider2):
+    return gjk_distance_original(collider1, collider2, return_interations=True)
```
Comment (AlexanderFabisch) [inline distance3d/gjk/_gjk_original.py:9]: We don't need a additional argument for that. You can just always return the number of iterations and cut it off here:

```python
return gjk_distance_original(collider1, collider2, return_interations=True)[:-1]
```

File: examples/visualizations/vis_robot_collision_objects.py [hunk 2]
```diff
@@ -143,8 +143,9 @@ def __call__(self, step, n_frames, tm, robot_bvh, world_bvh, joint_names):
     color = random_state.rand(3)
     box_artist = pv.Box(size=size, A2B=box2origin, c=color)
     box_artist.add_artist(fig)
-    box = colliders.Margin(
```
Comment (AlexanderFabisch) [inline examples/visualizations/vis_robot_collision_objects.py:146]: I think we can switch back to the original version.

**Comment by MaartenBehn** [comment]
Ready to merge.

When running an normal unit test the coverage is not 100% this is because it takes ages to run randomly take get every possible simplex configuration. I have added a function "test_compare_gjk_intersection_with_random_shapes_extensive()" that gets the coverage up to 100% but the function is commented out because it takes just too long and is not really relevant for a regular unit testing.

File: distance3d/colliders.py [hunk 3]
```diff
@@ -95,6 +95,26 @@ def aabb(self):
             Axis-aligned bounding box.
         """
 
+    @abc.abstractmethod
```
Comment (AlexanderFabisch) [inline distance3d/colliders.py:98]: This should be part of another pull request.

File: distance3d/colliders.py [hunk 4]
```diff
@@ -130,6 +150,20 @@ def update_pose(self, pose):
     def aabb(self):
         return np.array(axis_aligned_bounding_box(self.vertices)).T
 
+    def to_dict(self):
+        data = {
+            "typ": "ConvexHullVertices",
```
Comment (AlexanderFabisch) [inline distance3d/colliders.py:155]: Should be "type".

File: distance3d/colliders.py [hunk 5]
```diff
@@ -95,6 +95,26 @@ def aabb(self):
...
         """
 
+    @abc.abstractmethod
+    def to_dict(self):
+        """Return a data dict for json serialization
+
+        Returns
+        -------
+        data : dict
+            data dict
```
Comment (AlexanderFabisch) [inline distance3d/colliders.py:105]: Data does not say much...

File: distance3d/test/test_gjk_nesterov.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        shape2 = shape_names[random_state.randint(len(shape_names))]
+        args2 = random.RANDOM_GENERATORS[shape2](random_state)
+        collider1 = colliders.COLLIDERS[shape1](*args1)
+        collider2 = colliders.COLLIDERS[shape2](*args2)
+
+        intersection_libccd = gjk.gjk_intersection_libccd(collider1, collider2)
+        intersection_nesterov = gjk.gjk_nesterov_accelerated(collider1, collider2, use_nesterov_acceleration=False)[0]
+        intersection_nesterov_primitives = gjk.gjk_nesterov_accelerated_primitives(collider1, collider2, use_nesterov_acceleration=False)[0]
+
+        """
```
Comment (AlexanderFabisch) [inline distance3d/test/test_gjk_nesterov.py:31]: This should be removed

**Comment by AlexanderFabisch** [comment]
Unit tests are failing now when merging to master:

```
FAILED distance3d/test/test_gjk.py::test_gjk_boxes - AttributeError: 'Box' object has no attribute 'frame'
FAILED distance3d/test/test_gjk.py::test_gjk_spheres - AttributeError: 'Sphere' object has no attribute 'frame'
FAILED distance3d/test/test_gjk.py::test_gjk_cylinders - AttributeError: 'Cylinder' object has no attribute 'frame'
FAILED distance3d/test/test_gjk.py::test_gjk_capsules - AttributeError: 'Capsule' object has no attribute 'frame'
FAILED distance3d/test/test_gjk_nesterov.py::test_compare_gjk_intersection_flavours_with_random_shapes - AttributeError: 'Capsule' object has no attribute 'frame'
FAILED distance3d/test/test_gjk_nesterov.py::test_compare_gjk_intersection_flavours_with_random_shapes_with_nesterov_acceleration - AttributeError: 'Capsule' object has no attribute 'frame'
FAILED distance3d/test/test_gjk_nesterov.py::test_iterations - AttributeError: 'Sphere' object has no attribute 'frame'
FAILED distance3d/test/test_gjk_nesterov.py::test_to_many_iterations - AttributeError: 'Cylinder' object has no attribute 'frame'
FAILED distance3d/test/test_gjk_nesterov.py::test_to_far - AttributeError: 'Cylinder' object has no attribute 'frame'
FAILED distance3d/test/test_gjk_nesterov.py::test_compare_gjk_intersection_all_shapes - AttributeError: 'Capsule' object has no attribute 'frame'
FAILED distance3d/test/test_gjk_nesterov.py::test_compare_gjk_intersection_only_mesh - AttributeError: 'MeshGraph' object has no attribute 'frame'
```

**Comment by AlexanderFabisch** [comment]
@MaartenBehn Could you fix the unit tests?

File: distance3d/gjk/_gjk_original.py [hunk 7]
```diff
@@ -86,7 +90,7 @@ def gjk_distance_original(collider1, collider2):
                 else:
                     distance = new_solution.distance
 
-                return distance, closest_point1, closest_point2, simplex.points
+                return distance, closest_point1, closest_point2, simplex.points, iteration
```
Comment (AlexanderFabisch) [inline distance3d/gjk/_gjk_original.py:93]: This is not mentioned in the docstring

**Comment by AlexanderFabisch** [comment]
Unfortunately, two of the benchmarks fail:

```
python benchmarks/benchmark_gjk.py 
Original Mean: 0.11491; Std. dev.: 0.00319

Jolt:
Jolt intersection with Numba Mean: 0.00479; Std. dev.: 0.00552
Jolt intersection Mean: 0.00295; Std. dev.: 0.00005
Jolt distance with Numba Mean: 0.03174; Std. dev.: 0.00935
Jolt distance Mean: 0.02860; Std. dev.: 0.00008

Nesterov accelerated:
Nesterov Mean: 0.06174; Std. dev.: 0.01163
Nesterov with Numba Mean: 0.05728; Std. dev.: 0.00013
Nesterov with acceleration Mean: 0.08942; Std. dev.: 0.00027
Nesterov with acceleration and Numba Mean: 0.09004; Std. dev.: 0.00239

Nesterov accelerated Primitives:
Traceback (most recent call last):
  File "/home/dfki.uni-bremen.de/afabisch/Projekte/distance3d/benchmarks/benchmark_gjk.py", line 100, in <module>
    times = timeit.repeat(benchmark_nesterov_accelerated_primitives, repeat=10, number=1)
  File "/home/dfki.uni-bremen.de/afabisch/anaconda3/lib/python3.9/timeit.py", line 238, in repeat
    return Timer(stmt, setup, timer, globals).repeat(repeat, number)
  File "/home/dfki.uni-bremen.de/afabisch/anaconda3/lib/python3.9/timeit.py", line 205, in repeat
    t = self.timeit(number)
  File "/home/dfki.uni-bremen.de/afabisch/anaconda3/lib/python3.9/timeit.py", line 177, in timeit
    timing = self.inner(it, self.timer)
  File "<timeit-src>", line 6, in inner
  File "/home/dfki.uni-bremen.de/afabisch/Projekte/distance3d/benchmarks/benchmark_gjk.py", line 51, in benchmark_nesterov_accelerated_primitives
    gjk_nesterov_accelerated_primitives(shapes[i][0], shapes[i][1], use_nesterov_acceleration=False)
  File "/home/dfki.uni-bremen.de/afabisch/Projekte/distance3d/distance3d/gjk/_gjk_nesterov_accelerated_primitives.py", line 136, in gjk_nesterov_accelerated_primitives
    minkowski_diff = get_minkowski_diff(collider0, collider1)
  File "/home/dfki.uni-bremen.de/afabisch/Projekte/distance3d/distance3d/gjk/_gjk_nesterov_accelerated_primitives.py", line 99, in get_minkowski_diff
    data1, type1 = get_data_from_collider(collider1)
  File "/home/dfki.uni-bremen.de/afabisch/Projekte/distance3d/distance3d/gjk/_gjk_nesterov_accelerated_primitives.py", line 128, in get_data_from_collider
    assert type(collider) == Cylinder
AssertionError
```


```
python benchmarks/compare_gjk_iterations.py 
Case: 0
Traceback (most recent call last):
  File "/home/dfki.uni-bremen.de/afabisch/Projekte/distance3d/benchmarks/compare_gjk_iterations.py", line 27, in <module>
    collider1.round_values(6)
AttributeError: 'Cylinder' object has no attribute 'round_values'
```

**Comment by AlexanderFabisch** [comment]
@MaartenBehn

**Comment by AlexanderFabisch** [comment]
Merged, finally. :D  Thanks for the contribution!
